homeassistant/components/tado/services.py

```diff
@@ -0,0 +1,55 @@
+"""Services for the Tado integration."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers import selector
+
+from .const import (
+    ATTR_CONFIG_ENTRY,
+    ATTR_MESSAGE,
+    ATTR_READING,
+    DATA,
+    DOMAIN,
+    SERVICE_ADD_METER_READING,
+)
+
+_LOGGER = logging.getLogger(__name__)
+SCHEMA_ADD_METER_READING = vol.Schema(
+    {
+        vol.Required(ATTR_CONFIG_ENTRY): selector.ConfigEntrySelector(
+            {
+                "integration": DOMAIN,
+            }
+        ),
+        vol.Required(ATTR_READING, default=0): vol.Coerce(int),
+    }
+)
+
+
+@callback
+def setup_services(hass: HomeAssistant) -> None:
+    """Set up the services for the Tado integration."""
+
+    async def add_meter_reading(call: ServiceCall) -> None:
+        """Send meter reading to Tado."""
+        entry_id: str = call.data[ATTR_CONFIG_ENTRY]
+        reading: int = call.data[ATTR_READING]
+        _LOGGER.info("Add meter reading %s", reading)
+
+        tadoconnector = hass.data[DOMAIN][entry_id][DATA]
+        response: dict | None = await hass.async_add_executor_job(
+            tadoconnector.set_meter_reading, call.data[ATTR_READING]
+        )
+
+        if not response:
+            raise ServiceValidationError("Could not add meter reading")
+
+        if ATTR_MESSAGE in response:
```

Comment:
Why check against `ATTR_MESSAGE`. If there's a _valid response_, meaning the meter reading got added, it will return None. Right? If there's an exception, it will return the JSON reported back from Tado.

Reply:
`if not response: ...` is there because `tadoconnector.set_meter_reading` has a chance of returning `None` in case there is a `RequestError`. That is also why I suggested earlier to raise the `HomeAssistantError` instead of returning `None`.

`if ATTR_MESSAGE in response:` is there because the `message` is more descriptive then the `code`. Any error by Tado will return a dict with `message`. By checking only that `message` exists, I don't have to account for specific codes.

Reply:
My earlier suggestion was to raise `HomeAssistantError` in `tadoconnector.set_meter_reading` instead of returning `None` so that the connector (sometimes also called the coordinator) is coordinating the response. The service call will then only be responsible for raising `ServiceValidationError`

Reply:
Please refer to https://github.com/niro1987/homeassistant-core/pull/6 to review my suggestion above an rely feedback here before I commit the changes to this pr branch.

Reply:
I would be in favor of that. So go ahead. :)